osbrain/agent.py

```diff
@@ -1792,10 +1803,16 @@ def run_agent(name, nsaddr=None, addr=None, base=Agent, serializer=None,
     """
     if not nsaddr:
         nsaddr = os.environ.get('OSBRAIN_NAMESERVER_ADDRESS')
-    AgentProcess(name, nsaddr=nsaddr, addr=addr, base=base,
-                 serializer=serializer, transport=transport,
-                 attributes=attributes).start()
-    proxy = Proxy(name, nsaddr, safe=safe)
+    agent = AgentProcess(name=name, nsaddr=nsaddr, addr=addr, base=base,
+                         serializer=serializer, transport=transport,
+                         attributes=attributes)
+    agent_name = agent.start()
+    if name:
+        agent_name = name
+
+    print('AGENT NAME: ', name)
+
```

Comment:
I think you forgot this print and the `if name:`.

Reply:
Didn't forget that `if`. This one actually makes sense.

I did miss the `print`, though.

Reply:
Does the result change if you remove that `if`? (or does any test break?)

Reply:
It does change. This one is necessary, so the tests will probably fail if it's removed.

Reply:
Probably? Or for sure? :wink:

Reply:
Just tried it out, and they fail. So yeah, for sure 😉

Reply:
Which tests fail?

Reply:
`test_agent_auto_generated_name`,  which is the one I added for this.

This `if` avoids overriding the `agent_name` with the `name` passed as a parameter,  which is what we want to do if we didn't receive any parameters (basically the whole point of this PR). So yeah, it's important to keep it.

Reply:
I find that weird... Can you show me the `pytest` output so see where it fails? (or upload it to see the Travis report)

Reply:
How is that weird? I don't understand what you see wrong with these lines.
In the previous statement, we catch  the return value of `agent.start()`, but if a `name` was passed as a parameter, we should use that instead to create the `Proxy` on the next line. Otherwise we won't be able to get a `Proxy` from the nameserver because we will be querying for the wrong agent name.

Anyway, here is the full output from pytest (it times out while creating the object).
```
============================================== test session starts ==============================================
platform linux -- Python 3.6.4, pytest-3.3.2, py-1.5.2, pluggy-0.6.0 -- /usr/bin/python
cachedir: .cache
rootdir: /home/ocaballero/Documents/osbrain, inifile: setup.cfg
plugins: xdist-1.22.0, rerunfailures-4.0, pep8-1.0.6, mccabe-0.1, forked-0.2, flakes-2.0.0, cov-2.5.1
collected 258 items                                                                                             

osbrain/tests/test_agent.py::test_agent_auto_generated_name FAILED                                        [100%]

=================================================== FAILURES ====================================================
________________________________________ test_agent_auto_generated_name _________________________________________

nsproxy = <osbrain.proxy.NSProxy at 0x7faa3fc26940; not connected; for PYRONAME:Pyro.NameServer@127.0.0.1:11521>

    def test_agent_auto_generated_name(nsproxy):
        """
        If an agent is initialized without a name, a globally unique one must be
        auto generated for it.
        """
        count = 10
>       names = set([run_agent().get_attr('name') for x in range(count)])

osbrain/tests/test_agent.py:48: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
osbrain/tests/test_agent.py:48: in <listcomp>
    names = set([run_agent().get_attr('name') for x in range(count)])
osbrain/agent.py:1810: in run_agent
    proxy = Proxy(agent_name, nsaddr, safe=safe)
osbrain/proxy.py:80: in __init__
    while not self._ready_or_timeout(time0, timeout):
osbrain/proxy.py:96: in _ready_or_timeout
    self.unsafe.ping()
osbrain/proxy.py:154: in __getattr__
    return super().__getattr__(name)
/usr/lib/python3.6/site-packages/Pyro4/core.py:273: in __getattr__
    self._pyroGetMetadata()
/usr/lib/python3.6/site-packages/Pyro4/core.py:600: in _pyroGetMetadata
    self.__pyroCreateConnection()
/usr/lib/python3.6/site-packages/Pyro4/core.py:499: in __pyroCreateConnection
    uri = _resolve(self._pyroUri, self._pyroHmacKey)
/usr/lib/python3.6/site-packages/Pyro4/core.py:1853: in _resolve
    return nameserver.lookup(uri.object)
/usr/lib/python3.6/site-packages/Pyro4/core.py:187: in __call__
    return self.__send(self.__name, args, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Pyro4.core.Proxy at 0x7faa44621f28; not connected; for PYRO:Pyro.NameServer@127.0.0.1:11521>
methodname = 'lookup', vargs = ('@127.0.0.1:11521',), kwargs = {}, flags = 0, objectId = 'Pyro.NameServer'

    def _pyroInvoke(self, methodname, vargs, kwargs, flags=0, objectId=None):
        """perform the remote method call communication"""
        current_context.response_annotations = {}
        with self.__pyroConnLock:
            if self._pyroConnection is None:
                self.__pyroCreateConnection()
            serializer = util.get_serializer(self._pyroSerializer or config.SERIALIZER)
            objectId = objectId or self._pyroConnection.objectId
            annotations = self.__annotations()
            if vargs and isinstance(vargs[0], SerializedBlob):
                # special serialization of a 'blob' that stays serialized
                data, compressed, flags = self.__serializeBlobArgs(vargs, kwargs, annotations, flags, objectId, methodname, serializer)
            else:
                # normal serialization of the remote call
                data, compressed = serializer.serializeCall(objectId, methodname, vargs, kwargs, compress=config.COMPRESSION)
            if compressed:
                flags |= message.FLAGS_COMPRESSED
            if methodname in self._pyroOneway:
                flags |= message.FLAGS_ONEWAY
            self._pyroSeq = (self._pyroSeq + 1) & 0xffff
            msg = message.Message(message.MSG_INVOKE, data, serializer.serializer_id, flags, self._pyroSeq,
                                  annotations=annotations, hmac_key=self._pyroHmacKey)
            if config.LOGWIRE:
                _log_wiredata(log, "proxy wiredata sending", msg)
            try:
                self._pyroConnection.send(msg.to_bytes())
                del msg  # invite GC to collect the object, don't wait for out-of-scope
                if flags & message.FLAGS_ONEWAY:
                    return None  # oneway call, no response data
                else:
                    msg = message.Message.recv(self._pyroConnection, [message.MSG_RESULT], hmac_key=self._pyroHmacKey)
                    if config.LOGWIRE:
                        _log_wiredata(log, "proxy wiredata received", msg)
                    self.__pyroCheckSequence(msg.seq)
                    if msg.serializer_id != serializer.serializer_id:
                        error = "invalid serializer in response: %d" % msg.serializer_id
                        log.error(error)
                        raise errors.SerializeError(error)
                    if msg.annotations:
                        current_context.response_annotations = msg.annotations
                        self._pyroResponseAnnotations(msg.annotations, msg.type)
                    if self._pyroRawWireResponse:
                        msg.decompress_if_needed()
                        return msg
                    data = serializer.deserializeData(msg.data, compressed=msg.flags & message.FLAGS_COMPRESSED)
                    if msg.flags & message.FLAGS_ITEMSTREAMRESULT:
                        streamId = msg.annotations.get("STRM", b"").decode()
                        if not streamId:
                            raise errors.ProtocolError("result of call is an iterator, but the server is not configured to allow streaming")
                        return _StreamResultIterator(streamId, self)
                    if msg.flags & message.FLAGS_EXCEPTION:
                        if sys.platform == "cli":
                            util.fixIronPythonExceptionForPickle(data, False)
>                       raise data  # if you see this in your traceback, you should probably inspect the remote traceback as well
E                       Pyro4.errors.NamingError: unknown name: @127.0.0.1:11521

/usr/lib/python3.6/site-packages/Pyro4/core.py:472: NamingError
--------------------------------------------- Captured stdout setup ---------------------------------------------
Broadcast server running on 0.0.0.0:9091
NS running on 127.0.0.1:11521 (127.0.0.1)
URI = PYRO:Pyro.NameServer@127.0.0.1:11521
============================================= 257 tests deselected ==============================================
=================================== 1 failed, 257 deselected in 3.69 seconds ====================================
```

Reply:
From the traceback (seeing that your `proxy = Proxy(..)` is in line 1810), maybe you thought I was referring to change this:

```python
agent_name = agent.start()
if name:
    agent_name = name

proxy = Proxy(...)
```

Into this:

```python
agent_name = agent.start()
agent_name = name

proxy = Proxy(...)
```

But that does not make sense, as the first assignment would be unnecessary. I was referring to remove the if-clause all together:

```python
agent_name = agent.start()

proxy = Proxy(...)
```

Reply:
Yeah, I thought you meant that all along, hence my confusion 😁 

Removing the entire clause does work. When I wrote that I thought the agent name would only be returned if we passed nothing as  parameter. Now it makes sense to remove it 👍

Reply:
Candidate for the longest 2-lines discussion in GitHub. :joy: